File: javascript/src/museo/museoDemo.js
import * as Estados from "./estados.js";
import { fechaUTC, hoyUTC, sumarDias } from "./fechasUtil.js";
import { CatalogoObras } from "./catalogoObras.js";
import { Sala } from "./sala.js";
import { EncargadoCatalogo } from "./encargadoCatalogo.js";
import { Director } from "./director.js";
import { RestauradorJefe } from "./restauradorJefe.js";
import { Cuadro } from "./cuadro.js";
import { Escultura } from "./escultura.js";
import { OtroObjeto } from "./otroObjeto.js";
import { MonitorVestibulo } from "./monitorVestibulo.js";
import { Visitante } from "./visitante.js";
import { ProcesoRestauracionAutomatica } from "./procesoRestauracionAutomatica.js";
import { MuseoColaborador } from "./museoColaborador.js";

function fmtMoney(n) {
  return n.toLocaleString("en-US", { maximumFractionDigits: 0 });
}

function fmtDate(d) {
  if (d == null) return String(d);
  return d.toISOString().slice(0, 10);
}

function mostrarCatalogoObras(catalogo) {
  console.log();
  console.log("=== CATALOGO DE OBRAS ===");
  for (const obra of catalogo.listarObras()) {
    console.log(
      `ID ${obra.getIdObra()} | ${obra.getTitulo()} | ${obra.getAutor()} | ${obra.getPeriodo()} | Valor: $${fmtMoney(obra.getValorEconomico())} | Estado: ${obra.getEstado()}`
    );
    console.log(`    ${obra.mostrarDetalle()}`);
  }
}

function mostrarObrasEnSala(nombreSala, obras) {
  console.log();
  console.log(`--- Obras en sala: ${nombreSala} ---`);
  if (obras.length === 0) {
    console.log("  (ninguna)");
    return;
  }
  for (const obra of obras) {
    console.log(`  - ${obra.getTitulo()} - ${obra.getAutor()}`);
  }
}

function mostrarRestauracionesDeObra(obra, lista) {
  console.log();
  console.log(`--- Restauraciones de '${obra.getTitulo()}' (por antiguedad) ---`);
  if (lista.length === 0) {
    console.log("  (ninguna registrada)");
    return;
  }
  for (const r of lista) {
    const fin = r.getFechaFin() != null ? fmtDate(r.getFechaFin()) : "en curso";
    console.log(
      `  - #${r.getIdRestauracion()} | ${r.getTipoRestauracion()} | inicio ${fmtDate(r.getFechaInicio())} | fin ${fin} | motivo: ${r.getMotivo()}`
    );
  }
}

function main() {
  console.log();
  console.log("============================================================");
  console.log("SIMULACION: GESTION DEL MUSEO (JavaScript)");
  console.log("============================================================");

  const catalogo = new CatalogoObras();
  const salaRenacimiento = new Sala(1, "Renacimiento", "Planta 1 - Ala norte");
  const salaModerna = new Sala(2, "Arte moderno", "Planta 2");

  const encargado = new EncargadoCatalogo(
    1,
    "Ana Martinez",
    "ana@museo.org",
    true,
    "cat2024",
    "EC-001",
    catalogo
  );

  const director = new Director(
    2,
    "Luis Herrera",
    "director@museo.org",
    true,
    "dir2024",
    "Director general",
    catalogo
  );

  const restaurador = new RestauradorJefe(
    3,
    "Elena Vargas",
    "restauracion@museo.org",
    true,
    "res2024",
    "Pintura sobre lienzo"
  );

  console.log();
  console.log("=== INTENTO SIN AUTENTICACION (encargado del catalogo) ===");
  const cuadro = new Cuadro(
    1,
    "Noche estrellada",
    "Vincent van Gogh",
    "s. XIX",
    850_000_000.0,
    fechaUTC(1889, 6, 1),
    fechaUTC(2010, 3, 15),
    Estados.EXPUESTA,
    "oleo sobre lienzo",
    "postimpresionismo"
  );
  try {
    encargado.registrarObra(cuadro);
  } catch (e) {
    console.log(`Rechazado (correcto): ${e.message}`);
  }

  console.log();
  console.log("=== ENCARGADO DEL CATALOGO: sesion y registro de obras ===");
  encargado.iniciarSesion("cat2024");
  console.log(`Sesion iniciada: ${encargado.obtenerDatos()}`);

  const escultura = new Escultura(
    2,
    "El pensador",
    "Auguste Rodin",
    "s. XIX",
    420_000_000.0,
    fechaUTC(1902, 1, 1),
    fechaUTC(2005, 6, 1),
    Estados.EXPUESTA,
    "bronce",
    "modernista"
  );
  const otro = new OtroObjeto(
    3,
    "Sarcofago etrusco",
    "Anonimo",
    "s. III a.C.",
    95_000_000.0,
    fechaUTC(1750, 1, 1),
    fechaUTC(1998, 11, 20),
    Estados.EXPUESTA,
    "pieza arqueologica"
  );

  encargado.registrarObra(cuadro);
  encargado.registrarObra(escultura);
  encargado.registrarObra(otro);
  encargado.asignarSala(cuadro, salaRenacimiento);
  encargado.asignarSala(escultura, salaModerna);
  encargado.asignarSala(otro, salaRenacimiento);

  mostrarCatalogoObras(catalogo);
  console.log();
  console.log(`Catalogo actualizado el: ${fmtDate(catalogo.getFechaActualizacion())}`);

  console.log();
  console.log("=== VISITANTE: consulta en monitor del vestibulo ===");
  const salasMonitor = [salaRenacimiento, salaModerna];
  const monitor = new MonitorVestibulo(1, "Vestibulo principal", salasMonitor);
  const visitante = new Visitante("por_sala", monitor);
  mostrarObrasEnSala(salaRenacimiento.getNombre(), visitante.consultarObrasPorSala(salaRenacimiento.getIdSala()));
  mostrarObrasEnSala(salaModerna.getNombre(), visitante.consultarObrasPorSala(salaModerna.getIdSala()));

  console.log();
  console.log("=== PROCESO DIARIO: RESTAURACION AUTOMATICA (cada 5 anos) ===");
  const proceso = new ProcesoRestauracionAutomatica(catalogo);
  const hoy = hoyUTC();
  const candidatas = proceso.identificarObrasParaRestauracion(hoy);
  console.log(`Obras candidatas a revision periodica hoy (${fmtDate(hoy)}): ${candidatas.length}`);
  for (const o of candidatas) {
    console.log(
      `  - ID ${o.getIdObra()} - ${o.getTitulo()} (referencia ciclo desde ${fmtDate(o.fechaReferenciaProximoCicloRestauracion())})`
    );
  }

  console.log();
  console.log("=== RESTAURADOR JEFE: restauraciones ===");
  restaurador.iniciarSesion("res2024");
  cuadro.estado = Estados.DANADA;
  const restUrgente = restaurador.iniciarRestauracion(
    cuadro,
    "consolidacion de capa pictorica",
    hoy,
    "dano por humedad en sala (envio inmediato)"
  );
  console.log(
    `Restauracion urgente #${restUrgente.getIdRestauracion()} iniciada | obra en estado: ${cuadro.getEstado()}`
  );
  restaurador.finalizarRestauracion(restUrgente);
  console.log(
    `Restauracion #${restUrgente.getIdRestauracion()} finalizada | fin: ${restUrgente.getFechaFin() != null ? fmtDate(restUrgente.getFechaFin()) : null} | obra: ${cuadro.getEstado()}`
  );

  const restCiclo = restaurador.iniciarRestauracion(escultura, "limpieza y patina", hoy, "mantenimiento periodico");
  restaurador.finalizarRestauracion(restCiclo);

  mostrarRestauracionesDeObra(cuadro, restaurador.consultarRestauracionesPorObra(cuadro));
  console.log();
  console.log("--- Todas las restauraciones gestionadas por el jefe (por antiguedad) ---");
  for (const r of restaurador.consultarRestauraciones()) {
    console.log(
      `  - #${r.getIdRestauracion()} | obra ID ${r.getObra().getIdObra()} | ${r.getTipoRestauracion()} | ${fmtDate(r.getFechaInicio())}`
    );
  }

  console.log();
  console.log("=== DIRECTOR: museos colaboradores y cesiones ===");
  director.iniciarSesion("dir2024");
  const museoBogota = new MuseoColaborador(1, "Museo Nacional de Colombia", "Bogota", "Colombia");
  const museoMadrid = new MuseoColaborador(2, "Museo Thyssen-Bornemisza", "Madrid", "Espana");
  director.registrarMuseoColaborador(museoBogota);
  director.registrarMuseoColaborador(museoMadrid);
  console.log(`Museos colaboradores registrados: ${director.getMuseosColaboradores().length}`);

  const inicioCesion = hoyUTC();
  const finCesion = sumarDias(inicioCesion, 180);

  const cesion1 = director.cederObra(otro, museoBogota, inicioCesion, finCesion, 12_500_000.0);
  if (cesion1 == null) {
    throw new Error("cesion esperada");
  }
  console.log(
    `Cesion #${cesion1.getIdCesion()} | obra '${otro.getTitulo()}' -> ${museoBogota.getNombre()} | importe $${fmtMoney(cesion1.getImportePagado())} | estado obra: ${otro.getEstado()}`
  );

  const cesionEnCola = director.cederObra(otro, museoMadrid, inicioCesion, finCesion, 0.0);
  if (cesionEnCola == null) {
    console.log(`Segundo museo solicito la misma obra mientras sigue cedida: cola ${otro.colaSolicitudesCesion.length} pendiente(s).`);
  }

  const cesionSiguiente = director.finalizarCesionYAsignarSiguiente(cesion1);
  if (cesionSiguiente != null) {
    const c = cesionSiguiente;
    console.log(
      `Tras finalizar la primera cesion: nueva cesion #${c.getIdCesion()} -> ${c.getMuseo().getNombre()} | estado: ${c.getEstado()}`
    );
  } else {
    console.log("No hubo cesion siguiente (cola vacia o obra no disponible).");
  }

  console.log();
  console.log("=== VALORACION TOTAL DEL MUSEO ===");
  console.log(`Suma de valores de todas las obras: $${fmtMoney(director.consultarValorTotal())}`);

  encargado.cerrarSesion();
  director.cerrarSesion();
  restaurador.cerrarSesion();
  console.log();
  console.log("=== FIN DE LA SIMULACION ===");
  console.log();
}

main();

File: java/src/museo/MuseoDemo.java
package museo;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

/**
 * Demostracion en consola alineada con {@code python/museo.py} (main).
 */
public final class MuseoDemo {

    private MuseoDemo() {
    }

    public static void main(String[] args) {
        System.out.println();
        System.out.println("============================================================");
        System.out.println("SIMULACION: GESTION DEL MUSEO (Java)");
        System.out.println("============================================================");

        CatalogoObras catalogo = new CatalogoObras();
        Sala salaRenacimiento = new Sala(1, "Renacimiento", "Planta 1 - Ala norte");
        Sala salaModerna = new Sala(2, "Arte moderno", "Planta 2");

        EncargadoCatalogo encargado = new EncargadoCatalogo(
                1, "Ana Martinez", "ana@museo.org", true, "cat2024",
                "EC-001", catalogo
        );

        Director director = new Director(
                2, "Luis Herrera", "director@museo.org", true, "dir2024",
                "Director general", catalogo
        );

        RestauradorJefe restaurador = new RestauradorJefe(
                3, "Elena Vargas", "restauracion@museo.org", true, "res2024",
                "Pintura sobre lienzo"
        );

        System.out.println();
        System.out.println("=== INTENTO SIN AUTENTICACION (encargado del catalogo) ===");
        Cuadro cuadro = new Cuadro(
                1, "Noche estrellada", "Vincent van Gogh", "s. XIX",
                850_000_000.0,
                LocalDate.of(1889, 6, 1),
                LocalDate.of(2010, 3, 15),
                EstadosMuseo.EXPUESTA,
                "oleo sobre lienzo", "postimpresionismo"
        );
        try {
            encargado.registrarObra(cuadro);
        } catch (IllegalStateException e) {
            System.out.println("Rechazado (correcto): " + e.getMessage());
        }

        System.out.println();
        System.out.println("=== ENCARGADO DEL CATALOGO: sesion y registro de obras ===");
        encargado.iniciarSesion("cat2024");
        System.out.println("Sesion iniciada: " + encargado.obtenerDatos());

        Escultura escultura = new Escultura(
                2, "El pensador", "Auguste Rodin", "s. XIX",
                420_000_000.0,
                LocalDate.of(1902, 1, 1),
                LocalDate.of(2005, 6, 1),
                EstadosMuseo.EXPUESTA,
                "bronce", "modernista"
        );
        OtroObjeto otro = new OtroObjeto(
                3, "Sarcofago etrusco", "Anonimo", "s. III a.C.",
                95_000_000.0,
                LocalDate.of(1750, 1, 1),
                LocalDate.of(1998, 11, 20),
                EstadosMuseo.EXPUESTA,
                "pieza arqueologica"
        );

        encargado.registrarObra(cuadro);
        encargado.registrarObra(escultura);
        encargado.registrarObra(otro);
        encargado.asignarSala(cuadro, salaRenacimiento);
        encargado.asignarSala(escultura, salaModerna);
        encargado.asignarSala(otro, salaRenacimiento);

        mostrarCatalogoObras(catalogo);
        System.out.println();
        System.out.println("Catalogo actualizado el: " + catalogo.getFechaActualizacion());

        System.out.println();
        System.out.println("=== VISITANTE: consulta en monitor del vestibulo ===");
        ArrayList<Sala> salasMonitor = new ArrayList<>();
        salasMonitor.add(salaRenacimiento);
        salasMonitor.add(salaModerna);
        MonitorVestibulo monitor = new MonitorVestibulo(1, "Vestibulo principal", salasMonitor);
        Visitante visitante = new Visitante("por_sala", monitor);
        mostrarObrasEnSala(salaRenacimiento.getNombre(),
                visitante.consultarObrasPorSala(salaRenacimiento.getIdSala()));
        mostrarObrasEnSala(salaModerna.getNombre(),
                visitante.consultarObrasPorSala(salaModerna.getIdSala()));

        System.out.println();
        System.out.println("=== PROCESO DIARIO: RESTAURACION AUTOMATICA (cada 5 anos) ===");
        ProcesoRestauracionAutomatica proceso = new ProcesoRestauracionAutomatica(catalogo);
        List<ObraArte> candidatas = proceso.identificarObrasParaRestauracion();
        System.out.println("Obras candidatas a revision periodica hoy (" + LocalDate.now() + "): "
                + candidatas.size());
        for (ObraArte o : candidatas) {
            System.out.println("  - ID " + o.getIdObra() + " - " + o.getTitulo()
                    + " (referencia ciclo desde " + o.fechaReferenciaProximoCicloRestauracion() + ")");
        }

        System.out.println();
        System.out.println("=== RESTAURADOR JEFE: restauraciones ===");
        restaurador.iniciarSesion("res2024");
        cuadro.estado = EstadosMuseo.DANADA;
        Restauracion restUrgente = restaurador.iniciarRestauracion(
                cuadro, "consolidacion de capa pictorica", LocalDate.now(),
                "dano por humedad en sala (envio inmediato)"
        );
        System.out.println("Restauracion urgente #" + restUrgente.getIdRestauracion()
                + " iniciada | obra en estado: " + cuadro.getEstado());
        restaurador.finalizarRestauracion(restUrgente);
        System.out.println("Restauracion #" + restUrgente.getIdRestauracion() + " finalizada | fin: "
                + restUrgente.getFechaFin().orElse(null) + " | obra: " + cuadro.getEstado());

        Restauracion restCiclo = restaurador.iniciarRestauracion(
                escultura, "limpieza y patina", LocalDate.now(), "mantenimiento periodico"
        );
        restaurador.finalizarRestauracion(restCiclo);

        mostrarRestauracionesDeObra(cuadro, restaurador.consultarRestauracionesPorObra(cuadro));
        System.out.println();
        System.out.println("--- Todas las restauraciones gestionadas por el jefe (por antiguedad) ---");
        for (Restauracion r : restaurador.consultarRestauraciones()) {
            System.out.println("  - #" + r.getIdRestauracion() + " | obra ID " + r.getObra().getIdObra()
                    + " | " + r.getTipoRestauracion() + " | " + r.getFechaInicio());
        }

        System.out.println();
        System.out.println("=== DIRECTOR: museos colaboradores y cesiones ===");
        director.iniciarSesion("dir2024");
        MuseoColaborador museoBogota = new MuseoColaborador(1, "Museo Nacional de Colombia", "Bogota", "Colombia");
        MuseoColaborador museoMadrid = new MuseoColaborador(2, "Museo Thyssen-Bornemisza", "Madrid", "Espana");
        director.registrarMuseoColaborador(museoBogota);
        director.registrarMuseoColaborador(museoMadrid);
        System.out.println("Museos colaboradores registrados: " + director.getMuseosColaboradores().size());

        LocalDate inicioCesion = LocalDate.now();
        LocalDate finCesion = inicioCesion.plusDays(180);
        Optional<Cesion> optCesion1 = director.cederObra(otro, museoBogota, inicioCesion, finCesion, 12_500_000.0);
        if (!optCesion1.isPresent()) {
            throw new IllegalStateException("cesion esperada");
        }
        Cesion cesion1 = optCesion1.get();
        System.out.println("Cesion #" + cesion1.getIdCesion() + " | obra '" + otro.getTitulo()
                + "' -> " + museoBogota.getNombre()
                + " | importe $" + String.format("%,.0f", cesion1.getImportePagado())
                + " | estado obra: " + otro.getEstado());

        Optional<Cesion> cesionEnCola = director.cederObra(otro, museoMadrid, inicioCesion, finCesion, 0.0);
        if (!cesionEnCola.isPresent()) {
            System.out.println("Segundo museo solicito la misma obra mientras sigue cedida: cola "
                    + otro.colaSolicitudesCesion.size() + " pendiente(s).");
        }

        Optional<Cesion> cesionSiguiente = director.finalizarCesionYAsignarSiguiente(cesion1);
        if (cesionSiguiente.isPresent()) {
            Cesion c = cesionSiguiente.get();
            System.out.println("Tras finalizar la primera cesion: nueva cesion #" + c.getIdCesion()
                    + " -> " + c.getMuseo().getNombre() + " | estado: " + c.getEstado());
        } else {
            System.out.println("No hubo cesion siguiente (cola vacia o obra no disponible).");
        }

        System.out.println();
        System.out.println("=== VALORACION TOTAL DEL MUSEO ===");
        System.out.println("Suma de valores de todas las obras: $"
                + String.format("%,.0f", director.consultarValorTotal()));

        encargado.cerrarSesion();
        director.cerrarSesion();
        restaurador.cerrarSesion();
        System.out.println();
        System.out.println("=== FIN DE LA SIMULACION ===");
        System.out.println();
    }

    private static void mostrarCatalogoObras(CatalogoObras catalogo) {
        System.out.println();
        System.out.println("=== CATALOGO DE OBRAS ===");
        for (ObraArte obra : catalogo.listarObras()) {
            System.out.println("ID " + obra.getIdObra() + " | " + obra.getTitulo() + " | "
                    + obra.getAutor() + " | " + obra.getPeriodo()
                    + " | Valor: $" + String.format("%,.0f", obra.getValorEconomico())
                    + " | Estado: " + obra.getEstado());
            System.out.println("    " + obra.mostrarDetalle());
        }
    }

    private static void mostrarObrasEnSala(String nombreSala, List<ObraArte> obras) {
        System.out.println();
        System.out.println("--- Obras en sala: " + nombreSala + " ---");
        if (obras.isEmpty()) {
            System.out.println("  (ninguna)");
            return;
        }
        for (ObraArte obra : obras) {
            System.out.println("  - " + obra.getTitulo() + " - " + obra.getAutor());
        }
    }

    private static void mostrarRestauracionesDeObra(ObraArte obra, List<Restauracion> lista) {
        System.out.println();
        System.out.println("--- Restauraciones de '" + obra.getTitulo() + "' (por antiguedad) ---");
        if (lista.isEmpty()) {
            System.out.println("  (ninguna registrada)");
            return;
        }
        for (Restauracion r : lista) {
            String fin = r.getFechaFin().map(Object::toString).orElse("en curso");
            System.out.println("  - #" + r.getIdRestauracion() + " | " + r.getTipoRestauracion()
                    + " | inicio " + r.getFechaInicio() + " | fin " + fin + " | motivo: " + r.getMotivo());
        }
    }
}
